### PR TITLE
[#12048] Migrate FeedbackSessionClosingRemindersAction

### PR DESCRIPTION
**Pull request by EuniceSim142**
Part of #12048.

Per session, Closing-soon emails are sent for all co-owners, students and instructors (except students and instructors with deadline extensions).

---

For IT Tests, upon loading from data bundle, feedback sessions don't have association with deadline extension and feedback questions entities (ie: session.getDeadlineExtensions() and session.getFeedbackQuestions() return empty array). `prepareSession` function manually adds this association.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java [hunk 1]
```diff
@@ -39,18 +86,24 @@ public JsonResult execute() {
                 log.severe("Unexpected error", e);
             }
         }
+    }
 
+    private void executeForDatastorExtendedDeadlines() {
```
Comment (ziqing26) [inline src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:91]:
```suggestion
    private void executeForDatastoreExtendedDeadlines() {
```

File: src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java [hunk 2]
```diff
@@ -13,16 +13,63 @@
...
+            List<EmailWrapper> emailsToBeSent = sqlEmailGenerator.generateFeedbackSessionClosingEmails(session);
+            try {
+                taskQueuer.scheduleEmailsForSending(emailsToBeSent);
+                session.setClosingSoonEmailSent(true);
+            } catch (Exception e) {
+                log.severe("Unexpected error", e);
+            }
+        }
+
+        executeForDatastorExtendedDeadlines();
```
Comment (ziqing26) [inline src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:43]:
```suggestion
        executeForDatastoreExtendedDeadlines();
```

File: src/main/java/teammates/sqllogic/api/SqlEmailGenerator.java [hunk 3]
```diff
@@ -105,12 +105,12 @@ private List<EmailWrapper> generateFeedbackSessionOpeningOrClosingEmails(
...
             // student.
             students = students.stream()
-                    .filter(x -> userIds.contains(x.getId()))
+                    .filter(x -> !userIds.contains(x.getId()))
                     .collect(Collectors.toList());
 
             // instructor.
             instructors = instructors.stream()
-                    .filter(x -> userIds.contains(x.getId()))
+                    .filter(x -> !userIds.contains(x.getId()))
```
Comment (ziqing26) [inline src/main/java/teammates/sqllogic/api/SqlEmailGenerator.java:113]: Good catches

File: src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java [hunk 4]
```diff
@@ -13,16 +13,63 @@
...
  * Cron job: schedules feedback session closing emails to be sent.
  */
-class FeedbackSessionClosingRemindersAction extends AdminOnlyAction {
+public class FeedbackSessionClosingRemindersAction extends AdminOnlyAction {
 
     private static final Logger log = Logger.getLogger();
 
     @Override
     public JsonResult execute() {
+        executeForDatastoreFeedbackSessions();
```
Comment (ziqing26) [inline src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:28]: Just want to ask: Are the `executeForDatastore...` functions removable after data migration is fully done?

**Review by ziqing26 — CHANGES_REQUESTED** [review]
Thanks for the PR! Would it be possible to have one negative test case? The logic changes are generally sound with some small nitpicks

File: src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java (hunk 4 above)
Comment (EuniceSim142) [reply src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:28]: Yep, after migration can just easily remove the method definitions and calls

File: src/it/java/teammates/it/ui/webapi/FeedbackSessionClosingRemindersActionIT.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        MessageOutput response1 = (MessageOutput) actionOutput1.getOutput();
+
+        assertEquals("Successful", response1.getMessage());
+        assertTrue(session.isClosingSoonEmailSent());
+        assertTrue(de.isClosingSoonEmailSent());
+
+        // 7 email tasks queued:
+        // - 6 emails: 1 co-owner, 4 students and 3 instructors,
+        //             but 1 student and 1 instructor have deadline extensions (should not receive email)
+        // - 1 email:  1 student deadline extension
```
Comment (cedricongjh) [inline src/it/java/teammates/it/ui/webapi/FeedbackSessionClosingRemindersActionIT.java:155]: love these comments

File: src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java [hunk 6]
```diff
@@ -13,16 +13,63 @@
...
+            }
+        }
+
+        return new JsonResult("Successful");
+    }
+
+    private void executeForDatastoreFeedbackSessions() {
         List<FeedbackSessionAttributes> sessions = logic.getFeedbackSessionsClosingWithinTimeLimit();
 
         for (FeedbackSessionAttributes session : sessions) {
```
Comment (cedricongjh) [inline src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:75]: i think there might be a need for checking if the session has been migrated already

Comment (EuniceSim142) [reply src/main/java/teammates/ui/webapi/FeedbackSessionClosingRemindersAction.java:75]: Goode catch! 👍

**Review by cedricongjh — APPROVED** [review]
LGTM!

**Pull request merged**